### PR TITLE
Constant::eq skips spans

### LINKED ISSUE #77608: `mir::Constant::span` is a footgun for mir optimizations
The field is part of the `PartialEq` impl, and thus two constants that are equal except in their span will compare inequal. This causes problems in https://github.com/rust-lang/rust/pull/77551#discussion_r499759161 and https://github.com/rust-lang/rust/pull/77549#discussion_r499462146 and possibly in other places.

The "trivial" way to resolve this is to manually implement `PartialEq` and make it skip the `Span` field.
Downside: now we may cause weird diagnostics after such an optimization has been applied, because if one of the constants is eliminated, all diagnostics are reported on the other diagnostic's site. We'd need a multispan and properly combine it. That will be hard to ensure to happen if we just silently change `PartialEq`

I think it would be better to change [`mir::Body::required_consts`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/struct.Body.html#structfield.required_consts) from `Vec<mir::Constant>` to `FxHashMap<&'tcx ty::Const<'tcx>, Vec<Span>>` (or an `IndexMap` to preserve ordering for diagnostics) and remove the `span` field from `mir::Constant`.
This allows us to query the span of a constant when needed (we never need it for already evaluated constants as far as I can tell)

cc @rust-lang/wg-mir-opt

### PR DESCRIPTION
**Pull request by spastorino**
Opening this to start discussion and see what CI says about different approaches we can try for #77608.
For now I've just skipped span on `eq` implementation.

r? @oli-obk 

From what I've seen locally it passes all tests, which surprises me given the comments made on #77608. Although I'm not really sure how well this plays with incremental compilation, but my guess is that it doesn't play well.

Closes #77608

### CODE REVIEW

**Comment by oli-obk** [comment]
If you change PartialEq, you also need to adjust Hash and HashStable to reflect these changes, otherwise hashmaps behave badly

**Comment by Aaron1011** [comment]
I think this is going to cause problems with incremental compilation.  As @oli-obk said, this change would require ignoring the span in the `HashStable` impl. This means that a change in the span of a constant will not invalidate the result of a query, which could lead to stale diagnostics/debuginfo being produced (or even weirder issues if the `SyntaxContext` changes.

**Comment by spastorino** [comment]
yeah you both are right, I need to adjust that but I meant, wouldn't this cause incr. comp. issues regardless of the Hash and HashStable changes?.

File: compiler/rustc_middle/src/mir/terminator.rs [hunk 1]
```diff
@@ -527,7 +527,8 @@ impl<'tcx> TerminatorKind<'tcx> {
                         InlineAsmOperand::InOut { reg, late, in_value, out_place: None } => {
                             write!(fmt, "in{}out({}) {:?} => _", print_late(late), reg, in_value)?;
                         }
-                        InlineAsmOperand::Const { value } => {
+                        InlineAsmOperand::Const { value, .. } => {
+                            // FIXME consider span?
```
Comment (spastorino) [inline compiler/rustc_middle/src/mir/terminator.rs:531]: @oli-obk what should we do here? I think I wouldn't just do `write!(fmt, "span {:?}", span)` but unsure ...

File: compiler/rustc_mir/src/transform/const_prop.rs [hunk 2]
```diff
@@ -1085,7 +1083,8 @@ impl<'mir, 'tcx> MutVisitor<'tcx> for ConstPropagator<'mir, 'tcx> {
     fn visit_constant(&mut self, constant: &mut Constant<'tcx>, location: Location) {
         trace!("visit_constant: {:?}", constant);
         self.super_constant(constant, location);
-        self.eval_constant(constant, self.source_info.unwrap());
+        let source_info = self.source_info.unwrap();
+        self.eval_constant(source_info.span, constant, source_info);
```
Comment (spastorino) [inline compiler/rustc_mir/src/transform/const_prop.rs:1087]: There's a missing visit_span here

File: compiler/rustc_mir/src/util/pretty.rs [hunk 3]
```diff
@@ -438,14 +438,16 @@ fn use_verbose(ty: &&TyS<'tcx>) -> bool {
...
         match literal.ty().kind() {
             ty::Int(_) | ty::Uint(_) | ty::Bool | ty::Char => {}
             // Unit type
             ty::Tuple(tys) if tys.is_empty() => {}
             _ => {
                 self.push("mir::Constant");
-                self.push(&format!("+ span: {}", self.tcx.sess.source_map().span_to_string(*span)));
+                // FIXME implement visit_operand to take care of visit_span and visit_constant
+                // altogether
+                //self.push(&format!("+ span: {}", self.tcx.sess.source_map().span_to_string(*span)));
```
Comment (spastorino) [inline compiler/rustc_mir/src/util/pretty.rs:450]: This also needs to be fixed probably by implementing `visit_operand`

File: compiler/rustc_middle/src/mir/terminator.rs (hunk 1 above)
Comment (oli-obk) [reply compiler/rustc_middle/src/mir/terminator.rs:531]: hmm, yea, just print the span separately as you described

File: compiler/rustc_mir/src/interpret/eval_context.rs [hunk 4]
```diff
@@ -683,14 +683,13 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
         self.stack_mut().push(frame);
 
         // Make sure all the constants required by this frame evaluate successfully (post-monomorphization check).
-        for const_ in &body.required_consts {
-            let span = const_.span;
+        for (span, const_) in &body.required_consts {
```
Comment (oli-obk) [inline compiler/rustc_mir/src/interpret/eval_context.rs:686]: Oh right! we have required_consts. Perfect. I think we may be able to remove the span from Operand::Const after all

File: compiler/rustc_mir/src/transform/const_prop.rs [hunk 5]
```diff
@@ -474,8 +474,8 @@ impl<'mir, 'tcx> ConstPropagator<'mir, 'tcx> {
         match self.ecx.mir_const_to_op(&c.literal, None) {
             Ok(op) => Some(op),
             Err(error) => {
-                let tcx = self.ecx.tcx.at(c.span);
-                let err = ConstEvalErr::new(&self.ecx, error, Some(c.span));
+                let tcx = self.ecx.tcx.at(span);
+                let err = ConstEvalErr::new(&self.ecx, error, Some(span));
```
Comment (oli-obk) [inline compiler/rustc_mir/src/transform/const_prop.rs:478]: I wonder if we can use the span from the source_info instead

File: compiler/rustc_mir/src/transform/coverage/spans.rs [hunk 6]
```diff
@@ -865,9 +865,9 @@ pub(super) fn filtered_terminator_span(
         // Call `func` operand can have a more specific span when part of a chain of calls
         | TerminatorKind::Call { ref func, .. } => {
             let mut span = terminator.source_info.span;
-            if let mir::Operand::Constant(box constant) = func {
-                if constant.span.lo() > span.lo() {
-                    span = span.with_lo(constant.span.lo());
+            if let mir::Operand::Constant(box (c_span, _)) = func {
```
Comment (oli-obk) [inline compiler/rustc_mir/src/transform/coverage/spans.rs:868]: not sure why this special cased to constants. we can do the same thing for other operands by looking up their local's span. But this is essentially what I was talking about for pulling the span even outside the Operand and using (Operand, Span) tuples where necessary

File: compiler/rustc_mir_build/src/build/matches/test.rs [hunk 7]
```diff
@@ -429,17 +429,15 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
             block,
             source_info,
             TerminatorKind::Call {
-                func: Operand::Constant(box Constant {
-                    span: source_info.span,
-
+                func: Operand::Constant(box (source_info.span, Constant {
```
Comment (oli-obk) [inline compiler/rustc_mir_build/src/build/matches/test.rs:432]: here we could pull the span into a field on TerminatorKind::Call

**Comment by oli-obk** [comment]
the current changes should always stay their own commit, this is really clean

File: compiler/rustc_mir/src/interpret/eval_context.rs (hunk 4 above)
Comment (spastorino) [reply compiler/rustc_mir/src/interpret/eval_context.rs:686]: Should we do this in a follow up PR? or do you want me to take some kind of action here?

File: compiler/rustc_mir/src/transform/coverage/spans.rs (hunk 6 above)
Comment (spastorino) [reply compiler/rustc_mir/src/transform/coverage/spans.rs:868]: I'm not sure I follow your comment, are you referring to something pre-existing? in that case, should we address it in a follow up PR? or are you asking me to change how I'm handling spans further?

Comment (oli-obk) [reply compiler/rustc_mir/src/transform/coverage/spans.rs:868]: yea, I'm just musing on the pre-existing thing

**Comment by bstrie** [comment]
Triage, marking as experimental.

**Comment by Dylan-DPC** [comment]
closing this as it has bitrotted

**Pull request closed without merging**